probe-rs/src/architecture/arm/communication_interface.rs

```diff
@@ -52,34 +53,66 @@ pub trait ArmProbeInterface: DapAccess + SwoAccess + Debug + Send {
 
     fn read_from_rom_table(&mut self, dp: DpAddress) -> Result<Option<ArmChipInfo>, ProbeRsError>;
 
-    /// Deassert the target reset line
-    ///
-    /// When connecting under reset,
-    /// initial configuration is done with the reset line
-    /// asserted. After initial configuration is done, the
-    /// reset line can be deasserted using this method.
-    ///
-    /// See also [`Probe::target_reset_deassert`].
-    fn target_reset_deassert(&mut self) -> Result<(), ProbeRsError>;
-
     fn close(self: Box<Self>) -> Probe;
 }
 
-#[derive(Debug)]
-pub(crate) struct ArmCommunicationInterfaceState {
+// TODO: Rename trait!
+pub trait SwdSequence {
+    /// Corresponds to the DAP_SWJ_Sequence function from the ARM Debug sequences
+    fn swj_sequence(&mut self, bit_len: u8, bits: u64) -> Result<(), ProbeRsError>;
+
+    /// Corresponds to the DAP_SWJ_Pins function from the ARM Debug sequences
+    fn swj_pins(
+        &mut self,
+        pin_out: u32,
+        pin_select: u32,
+        pin_wait: u32,
+    ) -> Result<u32, ProbeRsError>;
+}
+
+pub trait UninitializedArmProbe: SwdSequence {
+    fn initialize(
+        self: Box<Self>,
+        sequence: Arc<dyn ArmDebugSequence>,
+    ) -> Result<Box<dyn ArmProbeInterface>, ProbeRsError>;
+
+    fn initialize_unspecified(self: Box<Self>) -> Result<Box<dyn ArmProbeInterface>, ProbeRsError> {
+        self.initialize(DefaultArmSequence::new())
+    }
+
+    /// Read DPDIR Register
+    fn read_dpidr(&mut self) -> Result<u32, ProbeRsError>;
+}
+
+pub trait ArmDebugState {}
+
+pub struct Uninitialized {
+    /// Specify if overrun detect should be enabled when the probe is initialized.
+    pub(crate) use_overrun_detect: bool,
+}
+
+pub struct Initialized {
     current_dp: Option<DpAddress>,
     dps: HashMap<DpAddress, DpState>,
+    use_overrun_detect: bool,
```

Comment:
This should still be used, the J-Link SWD implementation depends on this.

Reply:
Yep, agree :) I left this in this state intentionally so the warning would remind us about the open question about it. I think I can fix it now.